### PR TITLE
46 vertical window height

### PR DESCRIPTION
**Pull request by NicksPatties**
This PR is massive. Here's what I did:

- Wrote utility functions to find the current line, and organize exercise text and typed text into lines
- Implemented `renderText` function, which does what it says on the tin
- Wrote tests with as many edge cases as possible to verify code correctness
- Moved code related to rendering the exercise into the `exercise.go` file, added flags to support window size in `sweet.go`

TODO:
- enable vignetting, but I have implemented and tested that feature, so it should be straightforward.

### CODE REVIEW

File: cmd/root/exercise.go [hunk 1]
```diff
@@ -49,30 +45,155 @@ func (m exerciseModel) renderName() string {
...
+	n := '\n'
+	i := len(str) - 1
+	for ; i >= 0 && rune(str[i]) != n; i = i - 1 {
+	}
+
+	if i < 0 {
+		return str
+	}
+
+	return fmt.Sprintf("%s%s", str[:i], str[i+1:])
```
Comment (NicksPatties) [inline cmd/root/exercise.go:155]: return str[:i] + str[i+1:]?

File: cmd/root/exercise.go [hunk 2]
```diff
@@ -49,30 +45,155 @@ func (m exerciseModel) renderName() string {
...
+				currChar = renderVisibleRune(cursorStyle, c)
+			}
+			s += currChar
+		}
+		return
+	}
+
+	typed := *typedP
+
+	for i, exRune := range text {
```
Comment (NicksPatties) [inline cmd/root/exercise.go:187]: I could probably just redo this block as a expressionless switch statement, which may make things easier to read.

File: cmd/root/exercise_test.go [hunk 3]
```diff
@@ -6,10 +6,13 @@ import (
 
 	consts "github.com/NicksPatties/sweet/constants"
 	"github.com/NicksPatties/sweet/event"
+	"github.com/NicksPatties/sweet/util"
```
Comment (NicksPatties) [inline cmd/root/exercise_test.go:9]: newline

File: cmd/root/exercise_test.go [hunk 4]
```diff
@@ -82,17 +88,17 @@ func Test_renderText(t *testing.T) {
 		}
 		got := testModel.renderText()
 		if got != test.want {
-			t.Fatalf("%s failed\nexpected\n%s\ngot\n%s",
-				test.name, test.want, got,
+			t.Fatalf("%s failed\nwant\n%s\n%q\ngot\n%s\n%q\n",
```
Comment (NicksPatties) [inline cmd/root/exercise_test.go:91]: Fatalf or Errorf?

- Fatalf stops test output if there's an error. No other tests are run
- Errorf continues running tests, despite failures

I can control which tests are failing and which are succeeding by passing the test function to the `go test` command, if there's too much stuff going on. However, if I fix things, and then break other things, it's difficult to catch that earlier if I use Fatalf.

I should change these tests in this file to Errorf instead.

File: cmd/root/exercise_test.go [hunk 5]
```diff
@@ -104,24 +110,543 @@ func Test_renderText_cursorPosition(t *testing.T) {
...
+		if got != tc.want {
+			t.Fatalf("%s\ngot\n%v\n%s\nwant\n%v\n%s", tc.testName, got, renderBytes(got), tc.want, renderBytes(tc.want))
+		}
+	}
+}
+
+func Test_renderText_windowSize(t *testing.T) {
+	oldProfile := lg.ColorProfile()
+	lg.SetColorProfile(termenv.TrueColor)
+	defer lg.SetColorProfile(oldProfile)
```
Comment (NicksPatties) [inline cmd/root/exercise_test.go:231]: newline

File: cmd/root/exercise_test.go [hunk 6]
```diff
@@ -104,24 +110,543 @@ func Test_renderText_cursorPosition(t *testing.T) {
...
+		got := mockViewModel.renderText()
+
+		if got != tc.want {
+			t.Errorf("%s\ngot:\n%s\n%q\nwant:\n%s\n%q\n", tc.name, got, got, tc.want, tc.want)
+		}
+	}
+
+}
+
+func Test_renderLine(t *testing.T) {
```
Comment (NicksPatties) [inline cmd/root/exercise_test.go:422]: This block is irrelevant. these tests are already covered above

File: cmd/root/exercise_test.go [hunk 7]
```diff
@@ -104,24 +110,543 @@ func Test_renderText_cursorPosition(t *testing.T) {
...
+		typed string
+		want  int
+	}{
+		{
+			name:  "no typed text",
+			text:  "one\ntwo\nthree",
+			typed: "",
+			want:  0,
+		},
+		{
```
Comment (NicksPatties) [inline cmd/root/exercise_test.go:600]: Add one line but all newlines case.

File: cmd/root/sweet.go [hunk 8]
```diff
@@ -41,20 +41,24 @@ type viewOptions struct {
...
-	cursorStyle  lg.Style
-	typedStyle   lg.Style
-	mistakeStyle lg.Style
+	commentStyle         lg.Style
+	untypedStyle         lg.Style
+	cursorStyle          lg.Style
+	typedStyle           lg.Style
+	mistakeStyle         lg.Style
+	vignetteStyle        lg.Style
+	vignetteMistakeStyle lg.Style
```
Comment (NicksPatties) [inline cmd/root/sweet.go:50]: This property is no longer relevant, since I'm no longer vignetting the first line of the exercise
